api/envoy/config/route/v3/route_components.proto

```diff
@@ -1593,3 +1600,30 @@ message QueryParameterMatcher {
     bool present_match = 6;
   }
 }
+
+// HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+message InternalRedirectPolicy {
+  // An internal redirect is not handled, unless the number of previous internal redirects that a
+  // downstream request has encountered is lower than this value.
+  // In the case where a downstream request is bounced among multiple routes by internal redirect,
+  // the first route that hits this threshold, or does not set :ref:`internal_redirect_policy
+  // <envoy_api_field_config.route.v3.RouteAction.internal_redirect_policy>`
+  // will pass the redirect back to downstream.
+  //
+  // If not specified, at most one redirect will be followed.
+  google.protobuf.UInt32Value max_internal_redirects = 1;
+
+  // Defines what upstream response codes are allowed to trigger internal redirect. If unspecified,
+  // only 302 will be treated as internal redirect.
+  // Only 301, 302, 303, 307 and 308 are valid values. Any other codes will be ignored.
+  repeated uint32 redirect_response_codes = 2 [(validate.rules).repeated = {max_items: 5}];
+
+  // Specifies a list of predicates that are queried when an upstream response is deemed
+  // to trigger an internal redirect by all other criteria. Any predicate in the list can reject
+  // the redirect, causing the response to be proxied to downstream.
+  repeated google.protobuf.Any predicates = 3;
```

Comment:
Just a reminder that we want to switch this to a new message type, `TypedExtensionConfig` that includes a string and `Any`. This should live somewhere in `envoy.types` I think.

Reply:
Ack. I'll add that type. Just to confirm, it should be:
message TypedExtensionConfig {
  string name = 1;
  Any typed_config = 2;
}

Reply:
Yep. Maybe some comments too :) I'd point out in this that `name` can be arbitrary, and that `typed_config` should either be a concrete proto or of type https://github.com/cncf/udpa/blob/master/udpa/type/v1/typed_struct.proto

Reply:
I'm going to do this as an independent PR to unblock https://github.com/envoyproxy/envoy/pull/10553

Reply:
Ack. Sorry for not getting to this sooner. I am oncall this week.

Reply:
No worries, I have a PR out at https://github.com/envoyproxy/envoy/pull/11105

Reply:
This is done.